### PR TITLE
feat: Add asCommandActivity and asCommandResultActivity

### LINKED ISSUE #3560: port: Add AsCommandActivity and AsCommandResultActivity (#5452)
The changes in [Add AsCommandActivity and AsCommandResultActivity (#5452)](https://github.com/microsoft/botbuilder-dotnet/pull/5452) may need to be ported to maintain parity with `microsoft/botbuilder-dotnet`.

<blockquote>
Fixes https://github.com/microsoft/botbuilder-dotnet/issues/5451
</blockquote>

Please review and, if necessary, port the changes.

### PR DESCRIPTION
**Pull request by mdrichardson**
Fixes #3560

Adds these two methods to ActivityEx along with appropriate tests.

### Relevant .NET Links

* [ICommandActivity](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Schema/ICommandActivity.cs)
* [CommandValue](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Schema/CommandValue.cs)
* [ICommandResultActivity](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Schema/ICommandResultActivity.cs)
* [CommandResultValue](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Schema/CommandResultValue.cs)
* [Spec](https://github.com/microsoft/botframework-obi/blob/main/protocols/botframework-activity/botframework-activity.md#command-activity)

### CODE REVIEW

File: libraries/botframework-schema/src/activityEx.ts [hunk 1]
```diff
@@ -17,15 +17,18 @@ import {
...
+    ISuggestionActivity,
+} from './activityInterfaces';
 import {
     Activity,
     ActivityTypes,
     ChannelAccount,
-    ConversationReference, 
+    ConversationReference,
     Mention,
-    ResourceResponse,
```
Comment (mdrichardson) [inline libraries/botframework-schema/src/activityEx.ts:27]: this was unused. The rest of the changes up here are no-op.

File: libraries/botframework-schema/src/activityInterfaces.ts [hunk 2]
```diff
@@ -300,3 +300,61 @@ export interface ITraceActivity extends IActivity {
...
+export interface ICommandActivity<T> extends IActivity {
+    /**
+     * The name of the event.
+     */
+    name: string;
+    /**
+     * The value for this command.
+     */
+    value?: CommandValue<T>;
+}
```
Comment (mdrichardson) [inline libraries/botframework-schema/src/activityInterfaces.ts:331]: Note: I wasn't sure whether it was better to go with type generics here or something like `Record<string, unknown>`. Let me know if you want it changed.

Comment (joshgummersall) [reply libraries/botframework-schema/src/activityInterfaces.ts:331]: This makes sense - I wonder if there's some implicit meaning behind the fact that the .NET fields are JSON properties. This is a bit pedantic, but we can be even more strict with `T`. For example,

```typescript
type Foo<T extends number | string | boolean> = Promise<T>;
```

Informs the compiler that `T` must "be assignable to" one of the types in the union of `number | string | boolean`.

There's a good chance I'm over-engineering here, but it's something to consider even if it's just for the sake of learning about generics in Typescript 😄

File: libraries/botframework-schema/src/activityEx.ts [hunk 3]
```diff
@@ -292,6 +295,24 @@ export namespace ActivityEx {
         return isActivity(source, ActivityTypes.Handoff) ? source : null;
     }
 
+  /**
+   * Returns the source activity as an ICommandActivity object; or null, if this is not that type of activity.
+   * @param source The source activity.
+   * @returns This activity as a command activity; or null.
+   */
+     export function asCommandActivity<T>(source: Partial<Activity>): Partial<ICommandActivity<T>> {
```
Comment (joshgummersall) [inline libraries/botframework-schema/src/activityEx.ts:303]: Question - does the consumer of this file have to explicitly specify a type here? I.e.,

```typescript
ActivityEx.asCommandActivity<SomeType>(...)
```

In general, if the generic type can't be inferred by the compiler I tend to do `function<T = unknown>(...): T` as it makes the function usable without passing an explicit type.

Comment (mdrichardson) [reply libraries/botframework-schema/src/activityEx.ts]: I'm not sure. I wasn't able to find a spec for it. @EricDahlvang?

Switched all the type generics to use `<T = unknown>`, but maybe I should have only done the functions?

**Comment by joshgummersall** [comment]
@EricDahlvang would you give this a quick review? Just want to make sure the type semantic line up around the generic type parameters.

File: libraries/botframework-schema/src/activityInterfaces.ts (hunk 2 above)
Comment (EricDahlvang) [reply libraries/botframework-schema/src/activityInterfaces.ts]: Would it make sense to have ICommandActivity without T, and the value as any, then let consumers type it as they retrieve the .value?

**Review by joshgummersall — APPROVED** [review]

**Comment by mdrichardson** [comment]
@joshgummersall @stevengum After @joshgummersall merged `main` into this branch, the schema merge tests failed. So, I ran the tests locally to generate a new `tests.schema` file and committed it.

However, none of code in this PR deals with `tests.schema` and it looks like the schema merge tests all passed on previous commits to `main`. So, I have no idea why `tests.schema` is needing a change and this is throwing up a little bit of a red flag for me. Any idea what's going on?

**Comment by mdrichardson** [comment]
@joshgummersall @stevengum Disregard that last comment. This is due to a [change in sorting for BF-CLI](https://github.com/microsoft/botframework-cli/pull/1190). According to @chrimc62, this changed `tests.schema` should be merged.

**Pull request merged**